daemon/oci_linux.go

```diff
@@ -309,6 +309,12 @@ func setNamespaces(daemon *Daemon, s *specs.Spec, c *container.Container) error
 		s.Hostname = ""
 	}
 
+	// cgroup
+	if daemon.cgroupNamespacesEnabled && !c.HostConfig.Privileged {
```

Comment:
Wondering if this should be enabled by default if the host support it, or if it should be an opt-in/opt-out

I'm not too familiar with the feature itself, but my train of thought is; is this causing a change in behavior? Would there be situations where namespaced cgroups are not desirable?

@AkihiroSuda @crosbymichael ?

Reply:
I was on the fence about this myself - I listed a couple other options in [the GH issue](https://github.com/moby/moby/issues/38332).  @cpuguy83, any thoughts?

Reply:
The main change in behaviour is that `/proc/self/cgroup` will now hide the Docker-related stuff. Personally I think it should be enabled by default and shouldn't be daemon-wide at all -- it should be treated like all other namespaces. My only worry is that sharing a cgroup with `--cgroup-parent` will now become more complicated because if you have two containers in the same cgroup but in different cgroup namespaces then things start to get a bit weird (especially with `nsdelegate` thrown in).

Reply:
Sorry, I missed this notification.

Agree we should enable by default, but make sure not to break users who expect things like cgroup-parent and privileged to work a certain way.

Reply:
The current behaviour is to enable it by default (just like any other namespace), except if the container is privileged.

The main side effect (which shows up in the integration tests), is that the cgroup hierarchy is no longer visible from inside the container.  I don't know if this qualifies as a breaking change or not.

@cyphar / @cpuguy83, what are your concerns about running with `--cgroup-parent`?  I can add some integration tests to confirm the behaviour.

Reply:
> The main side effect (which shows up in the integration tests), is that the cgroup hierarchy is no longer visible from inside the container. I don't know if this qualifies as a breaking change or not.

This is the main purpose of cgroup namespaces, so it's good that it works. :wink: I don't think it'll count as a breaking change unless some folks have really tightly integrated into what `/proc/self/cgroups` contains which really isn't a good idea anyway.

> what are your concerns about running with `--cgroup-parent`?

Since writing the above comment I've remembered that cgroup sharing doesn't exist in Docker (getting two containers to share a cgroup) and so `--cgroup-parent` shouldn't be a problem after all. The concern would be that if you have two containers sharing a cgroup you would want them to share cgroup namespaces.

Reply:
We do support sharing of most namespaces, so we could add support for cgroup namespace sharing.